### ISSUE TITLE
Directional Focus and Window movement

### ISSUE THREAD

**Issue by VuiMuich** [leftwm/leftwm#771]
This topic is meant as reference for the regularly upcoming wish for directional movement of focus and windows.

As this seems to be mostly coming from new users who are not used to the simpler workflow yet I believe the priority for this is pretty low. But the effort to implement this is quite a bunch of areas of code to touch.
But we might have a poll eventually to see how many users would like to have this.

Now on some ideas how to approach this:
- Each window gets a `Neighbors` with directions `North`, `South`, `East`, `West` and a `Target: bool` fields 
- a method `FindNeighbors` runs on window creation and every command that changes window positions. Should it ignore floating windows? Should it find neighbors only in the current tag or across workspaces as well?
- if a window has multiple neighbors in one direction the following commands apply to the leftmost or topmost window
- we add commands `FocusNorth`, `FocusSouth`, `FocusEast`, `FocusWest` and `MoveTargetNorth`, `MoveTargetSouth`, `MoveTargetEast`, `MoveTargetWest`
- we add a new `border` type and color `target`
- target moving works as follows: target-mode is entered with the first `MoveTarget{}` command and stays active as long as the modifier(s) are kept pressed, we move the target border around and when all keys are released, we swap the current window and the target window
- `Sloppy` and `ClickTo` targeting modes might be nice too
- we might eventually add a target-mode to unfloat windows

I am a bit concerned that `FindNeighbors` might be somewhat expensive to calculate and therefore might make the affected actions sluggish, but that might just be my lack of knowledge of efficient algorithms for such tasks.

So far my thoughts, might extend that later.

**Comment by 0323pin** [comment]
@VuiMuich Forgive me if don't get it but, what's the advantage of this compared to `FocusWindowUp`/`Down`?

https://www.veed.io/view/49355471-e780-4e7e-9d45-bb30d71cda8f?sharingWidget=true

**Comment by VuiMuich** [comment]
I agree, that Focus Up/Down is completely fine!
It's just that this request comes up every now and then so having this issue to refer to when it comes up next time seemed useful to me.

And if that inspires someone to propose an implementation of this we can still decide if we accept it in `main` on top of the current behavior, or if we push it to `leftwm-contrib/patches` to keep the base implementation lean.

**Comment by hertg** [comment]
Isn't this a duplicate of #171?

**Comment by oredaze** [comment]
Maybe it is a good idea to reference the way this is done in other tiling wms. Here is a list of the ones I know have this functionality: i3, awesome, bspwm, herbstluftwm, xmonad, qtile.

As for the poll I believe it would be heavily biased, because people who are already using leftwm are people who most likely prefer navigating up or down a list, or else they would not be using leftwm.

Also making the navigation span across workspaces would be confusing. Same for it crossing between tiling and floating windows. Most likely would increase development effort too, so I advise against it.

**Comment by VuiMuich** [comment]
As this surfaced twice today and re-reading the stuff: I guess running `FindNeighbours` only on the directional commands might be sufficient and therefore wont affect overall performance if it turns out to be somewhat expensive. 
Though, if it is only calculating neighbours on the current tag, what could the realistic worst case scenario be? 30 Windows would even be insanely many windows and still probably been handled relatively quick, right?

Ignoring `floating` windows makes totally sense on one hand, as this really is about tiled windows. On the other hand gaining focus via keyboard might become awkward as you would need to keep the regular focus keybinds around as well.

**Comment by oredaze** [comment]
One important thing I would like to mention is that this type of focus potentially can behave badly if implemented badly. I mean the focusing algorithm for floating windows. You can have an arrangement where you intuitively would expect to focus a certain window, but instead you focus a different one, depending on the way it is implemented. I am fairly certain I have seen 3 or more **different** ways window managers handle it, and some of them were not intuitive. For example openbox is, but awesome a bit less I think. 

For the tiling part, maybe you want to think if remembering the last focused window is a good idea, if the direction of focus is ambiguous. For example if this: `|=` is 3 windows and you go from the left one to the right, which of the 2 right ones are you focusing? Top or bottom? They have equal priority in terms of geometry. Lets say you decide to always focus the top most one. But consider the case of having 3 or more on the right and you are using the bottom-most one, do you wanna go from the top to the bottom every time you focus right?

**Comment by 0323pin** [comment]
I'm sorry but, I still don't see the added value here.
It adds code complexity, potentially slowing down of the wm and IMHO unnecessary feature. Can these be implemented as feature that it's possible to disable?

**Comment by LGUG2Z** [comment]
[komorebi](https://github.com/LGUG2Z/komorebi) is also a list-based WM, but I managed to approximate some sort of directional focus and movement with some trial and error. If there is any interest someone could port [the relevant code](https://github.com/LGUG2Z/komorebi/blob/master/komorebi-core/src/direction.rs) to this project pretty easily I think.

**Comment by Mange** [comment]
> I'm sorry but, I still don't see the added value here.

The added value would be to be able to move focus around in the normal cardinal directions instead of treating windows as a 1D list.

It's currently {the thing,one of the things} that is preventing me from using leftWM right now.

> It adds code complexity, potentially slowing down of the wm and IMHO an unnecessary feature.

I don't understand how iterating a small list of windows to determine neighbors by their coordinates would be such a big deal.
Do you expect people to have 1000+ windows on a workspace, or is there some underlying complexity in leftwm that prevents this?

The idea is just to find the closest square to the north/left/south/east of another square. Game engines can do this millions of times per second. Here's some **pseudo code** written on a piece napkin that does something like it.

```rust
fn focus_right(focused: &Window, windows: &[Window]) -> Option<&Window> {
  let targets = windows
    .filter(|w| w != focused) // get other windows
    .filter(|w| w.coords.left >= focused.coords.right)  // only windows to the right
    .sort_by(|w| (w.coords.left - focused.coords.right)) // closest first
    .collect::Vec<_>();

  if targets.is_empty() {
    return None;
  }

  // limit to windows that are of the closest distance, in case windows to the right
  // are on a vertical stack
  let closest_x = targets[0].coords.left - focused.coords.right;

  // if multiple choices, just pick the one furthest up. Could be smarter in the future,
  // like picking the one closest in Y direction too.
  return targets.iter()
    .filter(|w| (w.coords.left - focused.coords.right) == closest_x)
    .min_by(|w| w.coords.y);
}
```

I understand that there's probably not just a global array of windows here, but there should be a way to iterate a list of windows and I also assume all windows must have information about the coordinates it is placed at somewhere (since they are laid out on the screen). (Oh, and I know the borrow checker would hate that code. Again, it's pseudo code.)

Yeah, this naive implementation will iterate the list of windows twice (worst case), but since `n` is so small here, does this really cause any problems of "slowness"? How large would `n` have to be for this to become an issue on a computer from 2005 or later?

**Comment by VuiMuich** [comment]
I would suggest to revisit this topic after #1000 getting merged.
But for now, let me put it like this:
- as interesting a project komorebi is, the referenced implementation imo shows exactly the point of "added code complexity" as there is a hardcoded case for every layout available, which is a horror to maintain if you want to support a flexible way to configure layouts, as intended in #1000 (@hertg please correct me if I am under a wrong impression here). Leave alone all the cases to consider like mixed tiled and floating as mentioned, or the then only logical desire to naviaget across workspaces etc.
- anyone wanting to contribute some possible solution is at all time welcome to open a PR proposing an implementation of such feature, just be aware of #1000 and its implications on changes regarding how layouts will be handled. (Adding to the expressed desire to make this feature optional at compile time, it would probably useful to put this in its own module.) All the regular contributors seem to be quite happy with one dimensional focus, so probably none of those would bi interested in investing time in this feature at this point.
- performance wise I am not as concerned any more then I was initially, though it still should not be underestimated. Just imagine a few clones here and there to make the borrowchecker happy and you end up copying large trees of data with each update cycle.

**Comment by marianomarciello** [comment]
Hello everyone, any news on this? I am be really interested in this feature. I am coming from bspwm/i3 where this feature is available and I use it really often. If you need some help with the implementation I will be available too.

**Comment by hertg** [comment]
Theoretically, we could introduce the feature in `leftwm-layouts` to find **which** `Rect` is top/right/bottom/left of a certain reference `Rect`. And then use this function from `leftwm`. Some pseudo-code from the top of my head.

In `leftwm-layouts`
```rust
// find the "rect" inside "rects" that is in the specified direction to the reference "rect"
fn find_positional(rects: [Rect], reference: usize, direction: Direction) -> usize {
   ...
}
```

In `leftwm`
```rust
// pseudo code
let rects = vec![...]; // a list of all rects
let reference_rect_index = get_rect_index_of_focused_window();
let direction = Direction::Right;
let next_rect_index = leftwm_layouts::find_positional(rects, focused_rect_index, direction);
let window = find_window_at_index(next_rect_index);
focus(window);
```

Might get more complicated than that with multi-monitor setups and cycling.

**Comment by VuiMuich** [comment]
Interesting proposal. It could be feature flagged in `lefwm-layouts` as well and `leftwm` uses this feature by default.

**Comment by VuiMuich** [comment]
> 
> 
> > * target moving works as follows: target-mode is entered with the first `MoveTarget{}` command and stays active as long as the modifier(s) are kept pressed, we move the target border around and when all keys are released, we swap the current window and the target window
> 
> 
> 
> Hi @VuiMuich I was reading this part of the first message in this issue. I think that moving the windows only when the keys are released could be not intuitive, since we don't have visual feedback until we release all the keys. In bspwm, i3 and other tiling window managers, the movement is done when the direction key is pressed. E.g. in i3 I have the following configuration:
> 
> ```
> 
> bindsym $mod+Shift+j move down
> 
> bindsym $mod+Shift+k move up
> 
> bindsym $mod+Shift+l move right
> 
> bindsym $mod+Shift+h move left
> 
> ```
> 
> So when I press $mod+Shift+j I have asp the windows swapped, Could it be better to follow this philosophy ? Am I misunderstanding your message ?
> 
> 
> 
> Thanks in advance for your time.
> 
> 

The idea of the target mode was inspired by the placement preview for new windows in `bspwm`. Adding some `instant_apply` option on top of this should be rather trivial in comparison to the rest of code needed.

**Comment by VuiMuich** [comment]
> Theoretically, we could introduce the feature in `leftwm-layouts` to find **which** `Rect` is top/right/bottom/left of a certain reference `Rect`. And then use this function from `leftwm`. Some pseudo-code from the top of my head.
> 
> 
> 
> In `leftwm-layouts`
> 
> ```rust
> 
> // find the "rect" inside "rects" that is in the specified direction to the reference "rect"
> 
> fn find_positional(rects: [Rect], reference: usize, direction: Direction) -> usize {
> 
>    ...
> 
> }
> 
> ```
> 
> 
> 
> In `leftwm`
> 
> ```rust
> 
> // pseudo code
> 
> let rects = vec![...]; // a list of all rects
> 
> let reference_rect_index = get_rect_index_of_focused_window();
> 
> let direction = Direction::Right;
> 
> let next_rect_index = leftwm_layouts::find_positional(rects, focused_rect_index, direction);
> 
> let window = find_window_at_index(next_rect_index);
> 
> focus(window);
> 
> ```
> 
> 
> 
> Might get more complicated than that with multi-monitor setups and cycling.

I like the idea of doing the positional calculations in `leftwm-layouts` for efficiency I'd still suggest to add a list of `window_neighbours` as a field to each window. This would hold window IDs for the `North`, `East`, `West` and `South` window and as suggested before for multiple neighbors prefer top and left (or leave this configurable).

**Comment by marianomarciello** [comment]
Hi @VuiMuich Thanks for your explanation. I have seen the PR #1000. Seems a huge PR so I will wait until the PR is done to start with the development of this feature.

**Comment by VuiMuich** [comment]
> Hi @VuiMuich Thanks for your explanation. I have seen the PR #1000. Seems a huge PR so I will wait until the PR is done to start with the development of this feature.

Sure.
If you are interested, this branch is basically stable enough for daily use and some live testing and feedback might actually be very welcome. Only thing one needs to be aware at this point is, that you need to wrap layoutnames in `"` in your config . We would like to get rid of this breakage, but its a bit tricky to come up with a solution.

**Comment by marianomarciello** [comment]
Hi @VuiMuich I saw that the #1000 PR has been merged.
I would like to start working on the Directional Focus implementation.
I would like to know if you use your main env to develop leftwm or do you have a vm ?
I don't know if a restart/reset action is implemented in leftwm or not yet.

Thanks for your time.

**Comment by Eskaan** [comment]
I use `make install-linked profile=dev` to create symlinks to the build directory and then just run `cargo build` followed by a reload (for me by keybind, you could of course also use `leftwm-command SoftReload`). The reload restarts everything but the `leftwm` binary itself (i.e. `leftwm-worker` gets restarted).

This should work most time when not breaking certain parts of the code that keep leftwm running (for example a panic would cause leftwm to go down).

If this happens I get dropped to the raw tty where I either
a. `vim` the file I edited and correct the mistake; then rebuild leftwm and `startx` again
b. If I can't figure out what went wrong, either stash my changes; rebuild and `startx` or manually link the leftwm binaries from the `target/release` folder so I can temporarily get a working gui.

I should mention that I also start my X11 session from a tty using `startx`.

Also a note in advance:
- When messing with the `State` struct, you can assume leftwm once won't be able to load the (now no longer matching) old state. You will primarily see that when all windows move to one tag after the reload.
- When changing the `Config` struct, you should also change your `config.ron` accordingly, otherwise you may fall back to the default config (also because it does no longer match i.e. missing fields).
- On a reload the core `leftwm` binary stays. That means some things like the core event loop won't be updated without a complete restart.

**Comment by marianomarciello** [comment]
Hello everyone again. Since this PR [22](https://github.com/leftwm/leftwm-layouts/pull/22) has been merged in leftwm-layouts. I think we can continue to develop this feature. I would like to take care of it, in my spare time. Since currently I am quite busy, this could take a while. :)

**Comment by marianomarciello** [comment]
Hello again everyone. I had time to do a part of this feature. Right now I am able only to focus windows in different direction. I have introduced 4 different command:
* FocusWindowNorth
* FocusWindowSouth
* FocusWindowEast
* FocusWindowWest

I would like to ask you if it's okay to introduce 4 new commands or is better to introduce only one command that takes as an input a direction. So something like:
* FocusDirWindow that take one of the following as parameter (North, South, East, West).

Beside of that, I think is better to divide those feature in two separte PR:
* One concerning only focus
* One concerning only window movement.

Let me know what do you think. :)

**Comment by mautamu** [comment]
Hey @marianomarciello!

Thanks for taking this on :). In my opinion, it would likely be best to use Focus/MoveWindowAt with the value for the direction ('E', 'N', 'S', etc.). Thinking this will allow us more flexibility without more complexity and (as much) duplicate code, but we can wait to see what others think.

Wondering if we also want to rename `FocusWindowUp`, `FocusWindowTop` and `FocusWindowDown` to, e.g. `FocusWindowBack`, `FocusWindowMain` and `FocusWindowForward` (as prevous/next is incorrect and up/down could be mistaken with north/south). 

> Beside of that, I think is better to divide those feature in two separate PR... movement.

Makes sense to me, thanks again for pushing the needle forward!

**Comment by marianomarciello** [comment]
Hey @mautamu 

Thanks for your time :). I think I will continue as you suggested using one function that takes as parameter one direction. After that I will start to do a PR so we can actually take a look at the code itself :). 

Beside that I would also like to implement the Directional Window movement feature, so I will open another issue and then create a similar API to move windows around in a specified direction :)
